shell/artifacts/hex/hex-game.js

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2018 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+defineParticle(({DomParticle}) => {
+  const rootTemplate = `
+    <style>
+      hex-game {
+        --hex-p1: #F44336;
+        --hex-p2: #2196F3;
+      }
+      hex-game[player=p1] {
+        --current-player-color: var(--hex-p1);
+      }
+      hex-game[player=p2] {
+        --current-player-color: var(--hex-p2);
+      }
+
+      /* TODO: Remove this once we switch slots to use shadow dom */
+      hexa-gon > div {
+        width: 100%;
+        height: 100%;
+      }
+      hex-cell {
+        width: 100%;
+        height: 100%;
+        display: block;
+      }
+      hex-cell:hover {
+        background: var(--current-player-color);
+        opacity: 0.5;
+      }
+      hex-cell[player=p1] {
+        background: var(--hex-p1);
+        opacity: 1;
+      }
+      hex-cell[player=p2] {
+        background: var(--hex-p2);
+        opacity: 1;
+      }
+    </style>
+    hex-game <hex-game player$="{{player}}" slotid="board"></hex-game>`;
+  const cellTemplate = `<hex-cell player$="{{player}}" on-click="onCellClick"></hex-cell>`;
+
+  class Board {
+    constructor(size) {
+      this._size = size;
+      this._board = [];
+      this._board.length = size * size;
+    }
+
+    get size() {
+      return this._size;
+    }
+
+    trySetCell(x, y, value) {
+      if (!this._board[x + this.size * y]) {
+        this._board[x + this.size * y] = value;
+        return true;
+      }
+      return false;
+    }
+
+    cell(x, y) {
+      return this._board[x + this.size * y];
+    }
+
+    toModel() {
+      let result = [];
+      for (let x = 0; x < this.size; x++) {
+        for (let y = 0; y < this.size; y++) {
+          let value = this.cell(x, y);
+          if (value) {
+            result.push({subId: `${x}-${y}`, player: value});
+          }
+        }
+      }
+      return result;
+    }
+  }
+
+  return class extends DomParticle {
+    constructor() {
+      super();
+      this._board = new Board(8);
+      this._player = 'p1';
+    }
+    getTemplate(slotName) {
+      if (slotName == 'root')
+        return rootTemplate;
+      else if (slotName == 'cell')
+        return cellTemplate;
+    }
+    render(props, state) {
+      if (this.currentSlotName == 'root') {
```

Comment:
Fwiw, if you simply returned both properties in one model, it would work in both cases. The template interpolation ignores unused model properties. Then again, perhaps Maria's code is intervening and doing something unexpected with the models to implement `consume set` ... this is the part that looks new to me.

Reply:
Done. This seems fine but I ran into a couple of issues: 1. the item model properties were being overwritten by the outer model properties (I flipped this around in dom-context). 2. I had a collision in naming: I had `player` in both model and item-model and was sometimes relying on item-model not being set. I worked around that by giving the item-model property a different name.